enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,295 @@
+	// +kubebuilder:validation:Required
+	// +required
+	Value string `json:"value"`
+}
+```
+
+
+
+### Implementation Details/Notes/Constraints
+A) Setting HTTP headers in the global section of haproxy.config via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds two new
+environment variables, `INGRESS_SET_HTTP_HEADERS`, which specify the headers to set in the 
+global section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `INGRESS_SET_HTTP_HEADERS` are comma-delimited lists of values of
+the form `name:value`.  (Colons and commas are not allowed in HTTP header
+names, so their use unambiguously delimits values in the environment variable.)
+The template translates these values into an appropriate `http-response set-header` stanzas.
+
+For example, if `ROUTER_SET_HTTP_HEADERS` is set to
+`X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block`, then the template adds the following stanzas to
+the appropriate frontends:
+
+```haproxy.config's global section
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting HTTP headers in the backend section `default_backend openshift_default` of the haproxy.config via Route's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+
+The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .
+The enhancement adds a field in ServiceAliasConfig called `RouteSetHTTPHeaders`, which specify the headers to set in the
+backend section `default_backend openshift_default` section of the haproxy.config.
+The router sets this variable based on the Route's
+`spec.httpSetHeaders` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
+
+The values of `RouteSetHTTPHeaders` are comma-delimited lists of values of
+the form `name:value`.  (Colons and commas are not allowed in HTTP header
+names, so their use unambiguously delimits values in the environment variable.)
+The template translates these values into an appropriate `http-response set-header` stanzas.
```

Comment:
Fix the number agreement:
```suggestion
The template translates these values into the appropriate `http-response set-header` stanzas.
```
Beyond that, you're repeating a sentence that you used in another section.  If you find yourself repeating some text, please either rework the text to avoid duplication or use [signposting](https://style.mla.org/effective-signposting/) to improve the cohesion of the text.  For example:
```suggestion
Similarly to the treatment of the `ROUTER_SET_HTTP_HEADERS` variable for global settings, the template translates the route-specific `RouteSetHTTPHeaders` values into the appropriate `http-response set-header` stanzas.
```

Reply:
this is no longer applicable. As from route we don't encode values.